crates/accelerate/src/commutation_analysis.rs

```diff
@@ -0,0 +1,197 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::PyModule;
+use pyo3::{pyfunction, pymodule, wrap_pyfunction, Bound, PyResult, Python};
+use qiskit_circuit::Qubit;
+
+use crate::commutation_checker::CommutationChecker;
+use hashbrown::HashMap;
+use indexmap::IndexSet;
+use pyo3::prelude::*;
+
+use pyo3::types::{PyDict, PyList};
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+// custom types: IndexSet allows to iterate over the elements in insertion
+// order, but uses the std hasher, which is slower than ahash
+type AIndexSet<T> = IndexSet<T, ::ahash::RandomState>;
+type CommutingNodes = Vec<AIndexSet<NodeIndex>>;
+
+/// Compute the commutation sets for a given DAG.
+///
+/// We return two HashMaps:
+///  * {wire: commutation_sets}: For each wire, we keep a vector of index sets, where each index
+///     set contains mutually commuting nodes. Note that these include the input and output nodes
+///     which do not commute with anything.
+///  * {(node, wire): index}: For each (node, wire) pair we store the index indicating in which
+///     commutation set the node appears on a given wire.
+///
+/// For example, if we have a circuit
+///
+///     |0> -- X -- SX -- Z (out)
+///      0     2    3     4   1   <-- node indices including input (0) and output (1) nodes
+///
+/// Then we would have
+///
+///     commutation_set = {0: [[0], [2, 3], [4], [1]]}
+///     node_indices = {(0, 0): 0, (1, 0): 3, (2, 0): 1, (3, 0): 1, (4, 0): 2}
+///
+#[allow(clippy::type_complexity)]
+fn analyze_commutations_inner(
+    py: Python,
+    dag: &mut DAGCircuit,
+    commutation_checker: &mut CommutationChecker,
+) -> PyResult<(
+    HashMap<Wire, CommutingNodes>,
+    HashMap<(NodeIndex, Wire), usize>,
+)> {
+    let mut commutation_set: HashMap<Wire, CommutingNodes> = HashMap::new();
+    let mut node_indices: HashMap<(NodeIndex, Wire), usize> = HashMap::new();
+
+    let max_num_qubits = 3;
```

Comment:
Not that it matters, but I'd have done this with:

```rust
const MAX_NUM_QUBITS: u32 = 3;
```

outside the function definition because it is never changed. But realistically the compiler will likely end up with the same generated code in both cases so it doesn't matter.

Reply:
Still consistency is nice -- I'll move it outside 👍🏻